Enhance `deno doc` result

afaur opened the pull request:
- Make sure names show up in `deno doc` command.
- Add mod entry points
  - Don't have to have these but..
    - `mod.ts` in `lib/util` simplifies `main.ts` and `mod.ts` by making them not change if files get added to `util`.
    - `mod.ts` in `lib` allows `main.ts` and `mod.ts` to have less duplication.
  - Maybe `main.ts` should be removed from `root` directory to use `mod.ts` as other Deno projects seem to do?
  - `main.ts` and `mod.ts` in the root directory are exactly the same right now.
    - If both are kept, maybe one should only export the `Main` object that contains `server` and `router`?
      - Doing this might provide a slimmer entry point (less imports), but I am not really sure how important that is.

**Update**:
  - Possible alterations (https://github.com/afaur/pogo/commit/248d692c1bcca8ea1b11986ab6465130dce17eca).
    - Rename `lib/mod.ts` to `lib/main.ts`
      - This file does not really function as a `mod` here.
      - This might be a good change to make.
    - Remove `main.ts` and fix `tests`
      - **Probably not a good decision since users of the library would have to change imports.**
    - Change `mod.ts` in the root to a one line export
      - `deno doc` stops showing the main entry point from the output with this change.
      - This might be okay.
        - The object/class that main provides, contains `server` and `router` functions.
          - Outside of the object/class they have their own entries in `deno doc`.

sholladay commented on main.ts:
```diff
@@ -1,12 +1,16 @@
+export * as serialize from './lib/serialize.ts';
+export * as server from './lib/server.ts';
+export * as toolkit from './lib/toolkit.ts';
+export * as types from './lib/types.ts';
```
Exporting the types like this is great, been meaning to do that. 👍

afaur commented:
There are some additional changes on this branch that are not currently showing up here. These additional changes remove the mod files, and add back in a semicolon that was left out. (see https://github.com/afaur/pogo/commit/b5cf751b22b42848de2117e48e1c163f47c2c88a)

sholladay commented on main.ts:
```diff
@@ -2,11 +2,18 @@ import Router, { RoutesList } from './lib/router.ts';
+export * as types from './lib/types.ts'
+
+export { default as Router } from './lib/router.ts';
+export { default as Server } from './lib/server.ts';
```
👍

sholladay commented on main.ts:
```diff
@@ -2,11 +2,18 @@ import Router, { RoutesList } from './lib/router.ts';
+export { default as Router } from './lib/router.ts';
+export { default as Server } from './lib/server.ts';
+
+export const Main = {
```
This implies `import { Main } from 'https://deno.land/x/pogo/main.ts'` is a fine pattern, but I would find that very unclear. How about instead of exporting a `Main` namespace, instead we provide named exports for the `server()` and `router()` factories and then also `export default` an object with them. I think the `deno doc` output should still be good, but correct me if I'm wrong.

afaur replied:
👍

sholladay commented on lib/toolkit.ts:
```diff
@@ -1,14 +1,10 @@
-}
-
-export default class Toolkit {
+export default class Toolkit implements ToolkitInterface {
```
I'm not sure I understand what value this change provides. How exactly does this change the `deno doc` output?

afaur replied:
It adds an `interface` under `namespace types` that describes `Toolkit`.
```sh
const router
const server
class Router
class Server
namespace types
  ...
  interface ToolkitInterface
  ...
```

afaur replied:
I added it so that the docs could explain what will be available on the `Toolkit`.
If you think it isn't necessary I can remove it.

sholladay replied:
I think it's great that it shows which helpers are available. The thing is, I thought `deno doc` was capable of describing the shape of a class without an interface. Maybe its not showing the methods because they lack TS Doc comments?

Also, is there a reason why you want to do it this way with `Toolkit`, but not `Router`, for example? Seems like whichever way we go on this, it should be consistent.

afaur replied:
- `deno doc` can describe the shape of a class, without an interface, so long as the class is exported from:
    - The `deno doc` entry point file (i.e. `main.ts`).
      -  Or whichever file is put in the deno doc badge on the readme.
  - I was putting interfaces in `types.ts` to:
    - Avoid re-exporting anything from `main.ts` that existed in `lib`
    - Avoid adding `jsdoc` to the class, and adding it to the `types` instead. (may not be what you prefer though)
    - The `jsdoc` could live on the `interface` as well as the `class`, but that may become burdensome. 
  - E.g. `jsdoc` entries will need to get added to the factories `server` and `router` inside of `main.ts`.

sholladay replied:
Ah, okay, I understand. You made it an interface because I said I didn't want to expose classes that users shouldn't instantiate themselves. That makes sense.

I guess I didn't think this all the way through. Having a seperate interface is probably more trouble than it's worth, just for the sake of documentation. I take back what I said earlier - let's expose the class, same as `Router`, etc.

sholladay commented on lib/types.ts:
```diff
@@ -2,6 +2,16 @@ import Request from './request.ts';
+    redirect(url: string): Response
+}
+
+export interface FileHandlerOptions {
```
👍

sholladay commented:
Sorry about the merge conflicts. I moved some things around while implementing `h.directory()`. If you could fix the conflict and remove the toolkit interface, in favor of exposing the class like you had it earlier, this will be ready to merge.

afaur commented:
No problem. I was able to merge this with master today, and the changes were easy to resolve.
I removed the interfaces that I added to `types`, and then added `tookit`, `request`, and `response` as exports from main.
I think this will allow for `jsdoc` strings on everything that is currently listed here (https://github.com/sholladay/pogo#api).

sholladay commented:
This is looking great. I'll try it out when I get to a terminal. Aiming to merge this tonight.
